### ISSUE TITLE
ffmpeg-git commit hash issue

### ISSUE THREAD

**Issue by Vangelis66** [jb-alvarado/media-autobuild_suite#277]
In the wee hours of the morning, my "non-free" ffmpeg compilation completed!

When I run my FFmpeg binary, the "banner" contains the following line:
`N-78561-g84eabb5`
The same info is printed when I run in the minnty console:
`$ cd /build/ffmpeg-git && sh version.sh`

The parts I sort of know are:

"N-78561" is a revision number (N tag?), although 
TBH I don't quite understand it... (what is e.g. N-0 or N-1?). 
If you can explain in layman's terms or point me to a website 
that does, I'll be much obliged... 

"g84eabb5" is short for "git-84eabb5" and is the 7-first-char 
abbreviation of the full SHA-1 string of the git commit.

I mostly use the pre-compiled free win32 Zeranoe builds: 
[win32/static/](http://ffmpeg.zeranoe.com/builds/win32/static/), 
a typical package name would be:
_ffmpeg-20160213-git-588e2e3-win32-static.7z_
If I visit: 
https://github.com/FFmpeg/FFmpeg/commits/master 
and search (in my browser) for `588e2e3`, I can locate 
the individual commit whose source code was used to build 
the downloaded package. For my Zeranoe example, it was: 
https://github.com/FFmpeg/FFmpeg/commit/588e2e338717078e88b0a7526e3740b7d3f6d207

Now, if I try to locate on the FFmpeg repo commit `84eabb5` 
(as displayed by my custom build) and limit search 
in the last, say, three days, I find nothing :-(

This issue vexed me quite a bit and spent the better part 
of the evening on finding answers on my own...
To cut a long story short, I was instructed by numerous 
google searches/answers to investigate contents of folder 
"/build/ffmpeg-git/.git/"
After... more reading, it turned out `84eabb5` is the 7-first-char of 
`$ cd /build/ffmpeg-git && cat .git/refs/heads/ab-suite`
while what I am truly after is: 
`$ cd /build/ffmpeg-git && cat .git/refs/remotes/origin/master` => 
`2e8ad2d65af575de2baca3b97dd646827bd4a0bc` => `2e8ad2d`
which finally leads me to: 
https://github.com/FFmpeg/FFmpeg/commit/2e8ad2d65af575de2baca3b97dd646827bd4a0bc , i.e.
**avcodec/mpeg12: Remove duplicate block_last_index setting code** 

If I have the standalone ffmpeg executable only (outside the confines of the suite) 
or if I recompile/delete "ffmpeg.git", how would `84eabb5` afford `2e8ad2d` ?
So, what I am asking as an "enhancement/feature request" is that - 
if at all possible - we mimic the Zeranoe site behaviour and have the final 
compiled ffmpeg binary print the commit hash info of the "remote" FFmpeg repo, 
not of the locally cloned "ab-suite" one (if that's what `84eabb5` is...). 
(Pardon my newbieness, still struggling to digest basic git terms...). 

Many, many thanks for this project!

**Comment by wiiaboo** [comment]
N-78561 is same as rev. 78561, yes

If patches are applied with `git am`they sort of ruin the git hash, so that's a disadvantage of applying them with it instead of `patch`. I prefer them because they're less troublesome to apply, create, update, clean, etc. For the OpenHEVC patch it'd have to be explicitely cleaned after compilation to not leave orphan files lying around. They can be applied with `patch` as well if you really need the version to be usable. It doesn't really help debugging, because you're not really using the revision as it's written.

**Comment by Vangelis66** [comment]
Hello...

I won't pretend I understood everything you said, 
but I think I got the gist of it:

1. For the suite to work, you need to apply locally 
two patches (ffmpeg-0001-blah.patch & ffmpeg-0002-blah.patch) 
on the sources fetched from ffmpeg-git.
2. You apply them by using a process called "[git am](https://www.kernel.org/pub/software/scm/git/docs/git-am.html)" 
because it's best for efficiency and maintenance reasons.
3. In doing so, the "remote" commit hash string gets changed, causing my "issue".
4. You could've used another process ("patch") to apply them that would not 
cause my "issue", but then that would add unneeded complexity and increase 
the maintenance toll of the suite code...

I have not yet updated the suite (still on git-1ea5dff from 20160214), 
so do not understand what "Append real version to version" actually does; 
is it something like `N-xxxxx-gxxxxxxx-gxxxxxxx`?
BTW, what is the best way to update? 
Is it safe to simply overwrite existing folders/files with the contents of 
unzipped "media-autobuild_suite-master.zip" ?
... And to see the effects of https://github.com/jb-alvarado/media-autobuild_suite/commit/112f98ac78356e3aff5949371f99cff457ff2a68 , should I first delete 
"/build/ffmpeg-git" or leave it in place and let the suite autoupdate it?

Indeed, very thankful for your help & guidance thus far!

**Comment by wiiaboo** [comment]
The version is going to be something like "ffmpeg version N-78592-g9070fe8-N-78590-g5590ab4" which is kinda ugly but does show what revision and commit it originally was before the patches.

Ideally, follow [these instructions](https://github.com/jb-alvarado/media-autobuild_suite/blob/master/doc/updating.md#using-git). If you don't feel like learning the basics of git, just replacing everything is fine too.

You shouldn't need to delete anything.

**Comment by Vangelis66** [comment]
@wiiaboo wrote: 

> The version is going to be something like "ffmpeg version N-78592-g9070fe8-N-78590-g5590ab4" 
> which is kinda ugly but does show what revision and commit it originally was before the patches.

My last compiled ffmpeg binary (on 22/02/2016) reports:

`ffmpeg version N-78635-g36ad681-g76377d6 Copyright (c) 2000-2016 the FFmpeg developers` 

so this has changed already; after looking at the commit history, 
it's due to https://github.com/jb-alvarado/media-autobuild_suite/commit/21b1f5971ff73d15a4850bd4b86eae11940b4942 . So, according to what you posted, 
the N-tag (revision) which refers to the original source should be 
78635 - 2 = 78633, hence my build is N-78633-g76377d6 
(not counting the extra 2 patches applied locally) ?

BTW, many thanks for providing an easy way to trace the 
original source from which the build is made of...
